File: Leetcode 125 - valid-palindrome/125.valid-palindrome.cpp
/*
 * @lc app=leetcode id=125 lang=cpp
 *
 * [125] Valid Palindrome
 */

// @lc code=start
#include <iostream>
#include <string>

using namespace std;

class Solution {
public:
  bool isPalindrome(string s) {
    int left = 0, right = s.size() - 1;

    while (left < right) {
      while (left < right && !isalnum(s[left])) {
        left++;
      }
      while (left < right && !isalnum(s[right])) {
        right--;
      }
      if (tolower(s[left]) != tolower(s[right])) {
        return false;
      }
      left++;
      right--;
    }

    return true;
  }
};

int main() {
  Solution sol;

  cout << boolalpha;
  string s = "A man, a plan, a canal: Panama";
  cout << sol.isPalindrome(s) << endl; // true

  string s2 = "race a car";
  cout << sol.isPalindrome(s2) << endl; // false

  return 0;
}
// @lc code=end

File: Leetcode 125 - valid-palindrome/125.valid-palindrome.java
/*
 * @lc app=leetcode id=125 lang=java
 *
 * [125] Valid Palindrome
 */

// @lc code=start
class Solution {

  public boolean isPalindrome(String s) {
    int left = 0, right = s.length() - 1;

    while (left < right) {
      // Skip non-alphanumeric characters from left
      while (left < right && !Character.isLetterOrDigit(s.charAt(left))) {
        left++;
      }

      // Skip non-alphanumeric characters from right
      while (left < right && !Character.isLetterOrDigit(s.charAt(right))) {
        right--;
      }

      // Compare characters ignoring case
      if (
        Character.toLowerCase(s.charAt(left)) !=
        Character.toLowerCase(s.charAt(right))
      ) {
        return false;
      }

      left++;
      right--;
    }

    return true;
  }

  public static void main(String[] args) {
    Solution sol = new Solution();

    System.out.println(sol.isPalindrome("A man, a plan, a canal: Panama")); // true
    System.out.println(sol.isPalindrome("race a car")); // false
    System.out.println(sol.isPalindrome(" ")); // true
  }
}
// @lc code=end
